Experimental - get rule failures with reason

neilbmclaughlin opened the pull request:
Bit late now but this was my attempt at an outline solution for returning informative failure information. If the approach looked good then the next step would have been to incorporate it in the results returned.

neilbmclaughlin committed:
Outline test

neilbmclaughlin committed:
Add rule description

neilbmclaughlin committed:
Simplify the returned data for failed rules to description and hint

djbingham commented on test/ffc-rules-engine.test.js:
```diff
@@ -669,3 +671,49 @@ describe('Rule: Water pollution reduction zone', () => {
+      ],
+      sssi: true
+    })
+    // expect.assertions(2)
```
Commented out code should be removed before merge.

djbingham reviewed (commented):
Looks like a nice implementation of informative messages. If this was to be continued, I guess messaging would be added to all the other rules and the tests expanded to cover that.

However, I think a good point was made yesterday about the rule definitions potentially being better placed in the Payment Orchestrator. Arguably, the Orchestrator should own the rule definitions, rather than fetching them from the rules engine only to pass them back in the `runEngine` requests.

neilbmclaughlin committed:
Switch from json to js format for rules

This was initially to allow the definition of an onSuccess and onFailure
function but, even though this is no longer being done, it seems less
jarring when switching between the code and rules.

neilbmclaughlin committed:
Add upper input bound to failed rule details

neilbmclaughlin committed:
Added lower input bound to failed rule details

Also, refactored tests for robustness

neilbmclaughlin committed:
Switch from adjustedPerimeter fact to adjustedPerimeterBounds

This now returns both upper and lower allowable limits.

neilbmclaughlin committed:
Refactor tests for clarity

neilbmclaughlin committed:
Refactor tests for clarity

neilbmclaughlin committed:
Add perimeter and area failure details

neilbmclaughlin committed:
Add adjusted area failure details

neilbmclaughlin committed:
Add lower bound checking

neilbmclaughlin committed:
Switch back to pond specific fact

neilbmclaughlin committed:
Add descriptions to remaining rules

neilbmclaughlin committed:
Move new run rules function into rules engine

neilbmclaughlin committed:
Switch tests over to runRules

Also, added new method someRulesPass
Need to remove runEngine. It's likely that runRules negates the needs to
return facts but need to establish that

neilbmclaughlin committed:
Refactor expectations to reduce duplication

neilbmclaughlin committed:
Partition tests by rules

neilbmclaughlin committed:
Organise schemas by directory

neilbmclaughlin committed:
Add sample for runRules

neilbmclaughlin committed:
More detail in example logging
